File: rd-core/src/test/kotlin/jetbrains/rd/util/test/cases/PropertyTest.kt
package com.jetbrains.rider.util.test.cases

import com.jetbrains.rider.util.Maybe
import com.jetbrains.rider.util.lifetime.Lifetime
import com.jetbrains.rider.util.lifetime.plusAssign
import com.jetbrains.rider.util.reactive.*
import org.testng.annotations.Test
import kotlin.test.assertEquals
import kotlin.test.assertFails
import kotlin.test.assertNotEquals

class PropertyTest {
    @Test
    fun testAdvise() {
        var acc = 0

        val property: IProperty<Int> = Property(acc)
        property.value = ++acc

        val log = arrayListOf<Int>()
        Lifetime.using { lifetime ->
            property.advise(lifetime, { x -> log.add(-x) })
            property.view(lifetime, { inner, x -> inner.bracket({ log.add(x) }, { log.add(10 + x) }) })

            lifetime += { log.add(0) }

            property.value = property.value
            property.value = ++acc
            property.value = ++acc
        }
        property.value = ++acc

        assertEquals(listOf(-1, 1, -2, 11, 2, -3, 12, 3, 0, 13), log)
    }

    @Test
    fun testWhenTrue() {
        var acc1 = 0
        var acc2 = 0

        val property: IProperty<Boolean> = Property(false)
        property.value = true
        Lifetime.using { lifetime ->
            property.whenTrue(lifetime, { lf -> acc1++ })
            property.whenTrue(lifetime, { it.bracket({ acc2 += 2 }, { acc2 -= 1 }) })
            assertEquals(1, acc1);assertEquals(2, acc2);

            property.value = true
            assertEquals(1, acc1);assertEquals(2, acc2);

            property.value = false
            assertEquals(1, acc1);assertEquals(1, acc2);

            property.value = true
            assertEquals(2, acc1);assertEquals(3, acc2);
        }
        assertEquals(2, acc1);assertEquals(2, acc2);
    }

    @Test
    fun testMaybe() {
        val property: IProperty<Boolean> = Property()
        assertEquals(Maybe.None, property.maybe)
        assertFails { property.value }

        var a = 0
        property.advise(Lifetime.Eternal, { a++ })
        assertEquals(0, a)

        property.set(false)
        assertNotEquals(Maybe.None, property.maybe)
        assertEquals(false, property.value)
        assertEquals(1, a)
    }

    @Test
    fun testBoolLogic() {
        val a: IProperty<Boolean> = Property()
        val b: IProperty<Boolean> = Property()

        val x = a.and(b)
        assertFails { x.value }

        a.set(false)
        assertFails { x.value }

        b.set(false)
        assertEquals(false, x.value)

        a.set(true)
        assertEquals(false, x.value)

        b.set(true)
        assertEquals(true, x.value)

        a.set(false)
        assertEquals(false, x.value)
    }

    @Test
    fun testBind() {
        val p = Property(0)
        val outer = Property<Int>()

        Lifetime.using { lf ->
            p.bind(lf, { outer.set(it) }, { setter -> outer.advise(lf, { setter(it) }) })

            assertEquals(0, outer.value)

            p.value = 1
            assertEquals(1, outer.value)

            outer.value = 2
            assertEquals(2, p.value)
        }

        p.value = 3
        assertEquals(2, outer.value)

        outer.value = 4
        assertEquals(3, p.value)

    }
}

File: rd-cpp/rd_core_cpp/src/test/cases/PropertyTest.cpp
//
// Created by jetbrains on 11.07.2018.
//

#include <gtest/gtest.h>

#include "Property.h"

TEST(property, advise) {
    int acc = 0;

    std::unique_ptr<IProperty<int> > property(new Property<int>(acc));

    property->set(++acc);

    std::vector<int> log;
    Lifetime::use<int>([&property, &acc, &log](Lifetime lifetime) {
        property->advise(lifetime, [&log](int const &x) {
            log.push_back(-x);
        });
        property->view(lifetime, [&log](Lifetime inner, int const &x) {
            inner->bracket(
                    [&log, &x]() { log.push_back(x); },
                    [&log, &x]() { log.push_back(10 + x); }
            );
        });

        lifetime->add_action([&log]() { log.push_back(0); });

        property->set(property->get());
        property->set(++acc);
        property->set(++acc);

        return 0;
    });

    EXPECT_EQ(log.size(), 10);

    property->set(++acc);

    std::vector<int> expected{-1, 1, 11, -2, 2, 12, -3, 3, 0, 13};
    EXPECT_EQ(expected, log);
}

TEST(property, when_true) {
    int acc1 = 0;
    int acc2 = 0;

    std::unique_ptr<IProperty<bool>> property(new Property<bool>(false));
    property->set(true);
    Lifetime::use<int>([&](Lifetime lifetime) {
        property->view(lifetime, [&acc1](Lifetime lt, bool const &flag) {
            if (flag) {
                acc1++;
            }
        });

        property->view(lifetime, [&](Lifetime lt, bool const &flag) {
            if (flag) {
                lt->bracket(
                        [&acc2]() { acc2 += 2; },
                        [&acc2]() { acc2 -= 1; }
                );
            }
        });

        EXPECT_EQ(1, acc1);
        EXPECT_EQ(2, acc2);

        property->set(true);
        EXPECT_EQ(1, acc1);
        EXPECT_EQ(2, acc2);

        property->set(false);
        EXPECT_EQ(1, acc1);
        EXPECT_EQ(1, acc2);

        property->set(true);
        EXPECT_EQ(2, acc1);
        EXPECT_EQ(3, acc2);

        return 0;
    });
    EXPECT_EQ(2, acc1);
    EXPECT_EQ(2, acc2);
}

TEST(property, view) {
    std::unique_ptr<IProperty<int>> property(new Property<int>(1));
    int save = 0;

    Lifetime::use<int>([&](Lifetime lifetime) {
        property->view(lifetime, [&](Lifetime lt, int const &value) {
            save = value;
        });
        property->set(2);
        EXPECT_EQ(2, save);
        lifetime->add_action([&property]() { property->set(-1); });
        EXPECT_EQ(2, save);

        return 0;
    });

    EXPECT_EQ(-1, property->get());
    EXPECT_EQ(-1, save);
}

TEST(property, uninitialized) {
    Property<int> property;

    EXPECT_THROW(property.get(), std::exception);

    std::vector<int> log;

    Lifetime::use([&](Lifetime lifetime) {
        property.advise(lifetime, [&log](int const &val) {
            log.push_back(val);
        });

        EXPECT_TRUE(log.empty());

        property.set(~0);

        EXPECT_EQ(~0, log[0]);
    });

    property.set(0);

    EXPECT_EQ(1, log.size());
}